main/src/ca/uwaterloo/flix/language/phase/PatternExhaustiveness.scala

```diff
@@ -0,0 +1,381 @@
+/*
+ * Copyright 2017 Jason Mittertreiner
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ca.uwaterloo.flix.language.phase
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.language.ast.Symbol.EnumSym
+import ca.uwaterloo.flix.language.ast.{Type, TypedAst}
+import ca.uwaterloo.flix.language.ast.TypedAst.Pattern
+import ca.uwaterloo.flix.language.ast.TypedAst.Pattern.Tag
+import ca.uwaterloo.flix.language.errors.ExhaustiveMatchError
+import ca.uwaterloo.flix.language.{CompilationError, GenSym}
+import ca.uwaterloo.flix.util.Validation
+import ca.uwaterloo.flix.util.Validation._
+
+/**
+  * The Pattern Exhaustiveness phase checks pattern matches for exhaustiveness
+  * as well as for useless patterns
+  *
+  * A pattern is useless if:
+  * A pattern match is exhaustive if:
+  */
+object PatternExhaustiveness extends Phase[TypedAst.Root, TypedAst.Root] {
+
+  /**
+    * Returns an error message if a pattern match is not exhaustive
+    */
+  def run(root: TypedAst.Root)(implicit flix: Flix): Validation[TypedAst.Root, CompilationError] = {
+    if (flix.options.debug) {
+      // Can probably do options stuff here
+    }
+    implicit val _ = flix.genSym
+    val defns = seqM(root.definitions.map { case (_, v) => Definition.CheckPats(root, v) })
+
+    defns.map ( _ => root)
+  }
+
+  object Definition {
+    def CheckPats(root: TypedAst.Root, tast: TypedAst.Declaration.BoundedLattice)(implicit genSym: GenSym): Validation[TypedAst.Declaration.BoundedLattice, CompilationError] = tast match {
+      case TypedAst.Declaration.BoundedLattice(tpe, bot, top, leq, lub, glb, loc) =>
```

Comment:
We use the naming convention that the "original" arguments have a zero, so `bot0`, `top0`, etc. This frees up the name `bot` for the "validated expression".